lua/lint/linters/clippy.lua

```diff
@@ -0,0 +1,51 @@
+local severities = {
+  note = vim.diagnostic.severity.INFO,
+  warning = vim.diagnostic.severity.WARN,
+  help = vim.diagnostic.severity.HINT,
+}
+
+local function parse(diagnostics, file_name, item)
+  for _, span in ipairs(item.spans) do
+    if span.file_name == file_name then
+      local message = item.message
+      if span.suggested_replacement ~= vim.NIL then
+        message = message .. " (" .. tostring(span.suggested_replacement) .. ")"
+      end
+
+      table.insert(diagnostics, {
+        lnum = span.line_start - 1,
+        end_lnum = span.line_end - 1,
+        col = span.column_start - 1,
+        end_col = span.column_end - 1,
+        severity = severities[item.level],
+        source = "clippy",
+        message = message
+      })
+    end
+  end
+
+  for _, child in ipairs(item.children) do
+    parse(diagnostics, file_name, child)
+  end
+end
+
+return {
+  cmd = "cargo",
+  args = { "clippy", "--message-format=json" },
+  stdin = false,
+  append_fname = false,
+  parser = function(output, bufnr)
+    local diagnostics = {}
+    local items = #output > 0 and vim.split(output, "\n") or {}
+    local file_name = vim.fn.expand("%")
```

Comment:
```suggestion
    local file_name = vim.api.nvim_buf_get_name(bufnr)
```

Given that nvim-lint runs the linter async by the time the output gets parsed the active buffer could've changed, so this should use the `bufnr` to resolve the name.